src/java.net.http/share/classes/jdk/internal/net/http/Stream.java

```diff
@@ -641,20 +641,35 @@ void handleReset(ResetFrame frame, Flow.Subscriber<?> subscriber) {
                 stateLock.unlock();
             }
             try {
-                int error = frame.getErrorCode();
-                IOException e = new IOException("Received RST_STREAM: "
-                        + ErrorFrame.stringForCode(error));
-                if (errorRef.compareAndSet(null, e)) {
-                    if (subscriber != null) {
-                        subscriber.onError(e);
+                final int error = frame.getErrorCode();
+                if (error == ErrorFrame.REFUSED_STREAM) {
+                    // A REFUSED_STREAM error code implies that the stream wasn't processed by the
+                    // peer and the client is free to retry the request afresh.
+                    // Here we arrange for the request to be retried.
+                    markUnprocessedByPeer();
+                    errorRef.compareAndSet(null, new IOException("request not processed by peer"));
+                    if (debug.on()) {
+                        debug.log("request unprocessed by peer (REFUSED_STREAM) " + this.request);
+                    }
```

Comment:
Hmm.... Should we just call:

```
    closeAsUnprocessed():
    return;
```

here?

Reply:
I've now updated this PR to follow this suggestion. The `finally` block which does `connection.decrementStreamsCount(streamid);` is needed (otherwise the `HttpClient` close() doesn't complete), so I haven't changed that part. 

With the latest state of this PR, a test-repeat 50 of java/net/httpclient tests continues to pass without any failures.